staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,339 @@
+	}
+	apiservers := apiserversi.(*sync.Map)
+	response := serviceableByResponse{}
+	var serviceableBy []string
+	apiservers.Range(func(key, value interface{}) bool {
+		apiserverKey := key.(string)
+		if apiserverKey == localAPIServerId {
+			response.locallyServiceable = true
+			return false
+		}
+		serviceableBy = append(serviceableBy, apiserverKey)
+		return true
+	})
+	response.serviceableBy = serviceableBy
+	return response, nil
+}
+
+func (h *peerProxyHandler) proxyRequestToDestinationAPIServer(req *http.Request, w http.ResponseWriter, hostname string) {
+	user, ok := apirequest.UserFrom(req.Context())
+	if !ok {
+		klog.Errorf("failed to get user info from request")
+		return
+	}
+
+	// write a new location based on the existing request pointed at the target service
+	location := &url.URL{}
+	location.Scheme = "https"
+	location.Host = hostname
+	location.Path = req.URL.Path
+	location.RawQuery = req.URL.Query().Encode()
+
+	newReq, cancelFn := apiserverproxyutil.NewRequestForProxy(location, req)
+	newReq.Header.Add(PeerProxiedHeader, "true")
+	defer cancelFn()
+
+	// create transport
+	clientConfig := &transport.Config{
+		TLS: transport.TLSConfig{
+			Insecure:   false,
+			CertFile:   h.proxyClientCertFile,
+			KeyFile:    h.proxyClientKeyFile,
+			CAFile:     h.peerCAFile,
+			ServerName: "kubernetes.default.svc",
+		},
+	}
+
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
+	proxyRoundTripper, transportBuildingError := transport.NewWithContext(ctx.Done(), clientConfig)
+	if transportBuildingError != nil {
+		klog.Error(transportBuildingError.Error())
+		return
+	}
+
+	proxyRoundTripper = transport.NewAuthProxyRoundTripper(user.GetName(), user.GetGroups(), nil, proxyRoundTripper)
+
+	handler := proxy.NewUpgradeAwareHandler(location, proxyRoundTripper, true, false, &responder{w: w})
+	handler.ServeHTTP(w, newReq)
+}
+
+// responder implements rest.Responder for assisting a connector in writing objects or errors.
+type responder struct {
+	w http.ResponseWriter
+}
+
+func (r *responder) Object(statusCode int, obj runtime.Object) {
+	responsewriters.WriteRawJSON(statusCode, obj, r.w)
+}
+
+func (r *responder) Error(w http.ResponseWriter, req *http.Request, err error) {
+	klog.Errorf("Error while proxying request to destination apiserver: %v", err)
+	http.Error(w, err.Error(), http.StatusServiceUnavailable)
+}
+
+func (h *peerProxyHandler) addSV(obj interface{}) {
+	sv := obj.(*v1alpha1.StorageVersion)
+	h.updateSVMap(nil, sv)
+}
+
+func (h *peerProxyHandler) updateSV(oldObj interface{}, newObj interface{}) {
+	oldSV := oldObj.(*v1alpha1.StorageVersion)
+	newSV := newObj.(*v1alpha1.StorageVersion)
+	h.updateSVMap(oldSV, newSV)
+}
+
+func (h *peerProxyHandler) deleteSV(obj interface{}) {
+	sv := obj.(*v1alpha1.StorageVersion)
+	h.updateSVMap(sv, nil)
+}
+
+func (h *peerProxyHandler) updateSVMap(oldSV *v1alpha1.StorageVersion, newSV *v1alpha1.StorageVersion) {
+
+	h.lock.Lock()
+	defer h.lock.Unlock()
```

Comment:
I think we should decide to use lock or sync.Map, not both, or I may be missing some other part that needs locking

Reply:
Yep, removed the lock since there was no need for it.